### ISSUE TITLE
concurrent map iteration and map write

### ISSUE THREAD

**Issue by rsanders** [sundy-li/burrowx#5]
I love burrowx - it's providing great visibility into our implementation of new streaming systems.

However, I have a regular crash.

```
2018/04/22 23:47:43 burrowx started,using server config:config/server.json, logfile cfg:/srv/burrowx/dev/config/logging.xml
2018/04/22 23:47:43 You could press [Ctrl+c] to stop burrowx
fatal error: concurrent map iteration and map write

goroutine 680 [running]:
runtime.throw(0x8654ad, 0x26)
        /usr/lib/golang/src/runtime/panic.go:596 +0x95 fp=0xc4206ba7b8 sp=0xc4206ba798
runtime.mapiternext(0xc4206bab00)
        /usr/lib/golang/src/runtime/hashmap.go:737 +0x7ee fp=0xc4206ba868 sp=0xc4206ba7b8
runtime.mapiterinit(0x7dbf00, 0xc42019ec60, 0xc4206bab00)
        /usr/lib/golang/src/runtime/hashmap.go:727 +0x2b3 fp=0xc4206ba8c0 sp=0xc4206ba868
github.com/sundy-li/burrowx/monitor.(*KafkaClient).CombineTopicAndConsumer(0xc42020c360)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:317 +0x58f fp=0xc4206bad00 sp=0xc4206ba8c0
github.com/sundy-li/burrowx/monitor.(*KafkaClient).getOffsets(0xc42020c360, 0xc4201d70e0, 0xc4206bafb0)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:282 +0x5ce fp=0xc4206baf88 sp=0xc4206bad00
github.com/sundy-li/burrowx/monitor.(*KafkaClient).Start.func3(0xc42020c360)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:156 +0x9f fp=0xc4206bafd8 sp=0xc4206baf88
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc4206bafe0 sp=0xc4206bafd8
created by github.com/sundy-li/burrowx/monitor.(*KafkaClient).Start
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:158 +0x12e
```

Config file and log file attached.

[burrowx-crash-log.txt](https://github.com/sundy-li/burrowx/files/1953519/burrowx-crash-log.txt)

[server.json.txt](https://github.com/sundy-li/burrowx/files/1953520/server.json.txt)

**Comment by sundy-li** [comment]
Thanks for your report, I just fix this bug in latest code, it was caused by ` concurrent map read and write `. So I simply add a lock to fix it.

**Comment by rsanders** [comment]
I pulled your latest code, and it looks like it's still happening.

```
2018/04/29 02:38:30 You could press [Ctrl+c] to stop burrowx
fatal error: concurrent map iteration and map write

goroutine 77 [running]:
runtime.throw(0x8657ed, 0x26)
        /usr/lib/golang/src/runtime/panic.go:596 +0x95 fp=0xc42051a758 sp=0xc42051a738
runtime.mapiternext(0xc42051a908)
        /usr/lib/golang/src/runtime/hashmap.go:737 +0x7ee fp=0xc42051a808 sp=0xc42051a758
github.com/sundy-li/burrowx/monitor.(*KafkaClient).CombineTopicAndConsumer.func1()
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:318 +0x323 fp=0xc42051a9d8 sp=0xc42051a808
github.com/sundy-li/burrowx/monitor.withReadLock(0x0, 0x0, 0x0, 0xc42051ab08)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:491 +0x85 fp=0xc42051aa08 sp=0xc42051a9d8
github.com/sundy-li/burrowx/monitor.(*KafkaClient).CombineTopicAndConsumer(0xc4201ce000)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:352 +0x466 fp=0xc42051ad00 sp=0xc42051aa08
github.com/sundy-li/burrowx/monitor.(*KafkaClient).getOffsets(0xc4201ce000, 0xc42016d560, 0xc42051afb0)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:282 +0x5ce fp=0xc42051af88 sp=0xc42051ad00
github.com/sundy-li/burrowx/monitor.(*KafkaClient).Start.func3(0xc4201ce000)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:156 +0x9f fp=0xc42051afd8 sp=0xc42051af88
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc42051afe0 sp=0xc42051afd8
created by github.com/sundy-li/burrowx/monitor.(*KafkaClient).Start
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:158 +0x12e
```

It looks like perhaps the consumerOffsetMapLock should also be held there?

```

**Comment by sundy-li** [comment]
Sorry, please pull the latest code again, hope this bug is fixed now  :)

**Comment by rsanders** [comment]
Still seeing it, unfortunately.

```

fatal error: concurrent map iteration and map write

goroutine 432 [running]:
runtime.throw(0x86598d, 0x26)
        /usr/lib/golang/src/runtime/panic.go:596 +0x95 fp=0xc42093c760 sp=0xc42093c740
runtime.mapiternext(0xc42093c910)
        /usr/lib/golang/src/runtime/hashmap.go:737 +0x7ee fp=0xc42093c810 sp=0xc42093c760
github.com/sundy-li/burrowx/monitor.(*KafkaClient).CombineTopicAndConsumer.func1()
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:317 +0x323 fp=0xc42093c9e0 sp=0xc42093c810
github.com/sundy-li/burrowx/monitor.withReadLock(0xc420157960, 0xc42093cb08)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:490 +0x5b fp=0xc42093ca08 sp=0xc42093c9e0
github.com/sundy-li/burrowx/monitor.(*KafkaClient).CombineTopicAndConsumer(0xc42025e460)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:351 +0x45a fp=0xc42093cd00 sp=0xc42093ca08
github.com/sundy-li/burrowx/monitor.(*KafkaClient).getOffsets(0xc42025e460, 0xc420213500, 0xc42093cfb0)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:282 +0x5ce fp=0xc42093cf88 sp=0xc42093cd00
github.com/sundy-li/burrowx/monitor.(*KafkaClient).Start.func3(0xc42025e460)
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:156 +0x9f fp=0xc42093cfd8 sp=0xc42093cf88
runtime.goexit()
        /usr/lib/golang/src/runtime/asm_amd64.s:2197 +0x1 fp=0xc42093cfe0 sp=0xc42093cfd8
created by github.com/sundy-li/burrowx/monitor.(*KafkaClient).Start
        /home/rsanders/go/src/github.com/sundy-li/burrowx/monitor/client.go:158 +0x12e

```

**Comment by sundy-li** [comment]
@rsanders  Sorry, for now on I had not enough time to make a test. I just notice the bug and fixed that. Hope you could help me test with that.

I would like to use sync.Map, it will make much better,  but it requires Go 1.9+.

Burrowx needs 4 maps to store the kafka state(topics, consumers, partitions, brokers), so it was a little complicated to make the concurrent , Maybe I should make the code much cleaner :)

**Comment by rsanders** [comment]
Your latest commit fixed this problem for me - been running for 2.5 days without a crash.  Thank you!
